### ISSUE TITLE
ShuffledDataSourcesArrowExamplesIterable cannot properly resume from checkpoint

### ISSUE THREAD

**Issue by howitry** [huggingface/datasets#7901]
### Describe the bug

ShuffledDataSourcesArrowExamplesIterable cannot properly resume from checkpoint

### Steps to reproduce the bug

1. The reproducible code is as follows：

```
from datasets import Dataset, concatenate_datasets, interleave_datasets
ds = Dataset.from_dict({"a": range(12)}).to_iterable_dataset(num_shards=1)
ds = ds.shuffle(seed=42)

for idx, example in enumerate(ds):
    print(example)
    if idx == 2: #The checkpoint can be loaded correctly only when idx <= 1.
        state_dict = ds.state_dict()
        print("checkpoint")
        break

print("state_dict: ",state_dict)

ds.load_state_dict(state_dict)
print(f"restart from checkpoint")
for example in ds:
    print(example)
```
2. The error message is as follows:
```
{'a': 0}
{'a': 7}
{'a': 6}
checkpoint
state_dict:  {'examples_iterable': {'examples_iterable': {'examples_iterable': {'shard_idx': 1, 'shard_example_idx': 0, 'type': 'ShuffledDataSourcesArrowExamplesIterable'}, 'previous_state': {'shard_idx': 0, 'shard_example_idx': 0, 'type': 'ShuffledDataSourcesArrowExamplesIterable'}, 'batch_idx': 12, 'num_chunks_since_previous_state': 12, 'cropped_chunk_length': 0, 'type': 'RebatchedArrowExamplesIterable'}, 'previous_state': {'examples_iterable': {'shard_idx': 1, 'shard_example_idx': 0, 'type': 'ShuffledDataSourcesArrowExamplesIterable'}, 'previous_state': {'shard_idx': 0, 'shard_example_idx': 0, 'type': 'ShuffledDataSourcesArrowExamplesIterable'}, 'batch_idx': 12, 'num_chunks_since_previous_state': 12, 'cropped_chunk_length': 0, 'type': 'RebatchedArrowExamplesIterable'}, 'batch_idx': 3, 'num_chunks_since_previous_state': 2, 'cropped_chunk_length': 0, 'type': 'RebatchedArrowExamplesIterable'}, 'epoch': 0}
restart from checkpoint
Loading a state dict of a shuffle buffer of a dataset without the buffer content.The shuffle buffer will be refilled before starting to yield new examples.
```
### Expected behavior

I want a correct resume from any checkpoint, but currently the checkpoint can only be loaded correctly when idx <= 1.

### Environment info

datasets Version: 4.4.1

@lhoestq

**Comment by lhoestq** [comment]
Hi ! As you can read in the logs, the shuffle buffer content is lost when resuming a shuffled dataset. The default size is 1000 examples, but you can tweak it

e.g. if you run your code with this

```diff
- ds = Dataset.from_dict({"a": range(12)}).to_iterable_dataset(num_shards=1)
- ds = ds.shuffle(seed=42)
+ ds = Dataset.from_dict({"a": range(100)}).to_iterable_dataset(num_shards=1)
+ ds = ds.shuffle(seed=42, buffer_size=10)
```

then you get

```
{'a': 0}
{'a': 7}
{'a': 6}
checkpoint
restart from checkpoint
Loading a state dict of a shuffle buffer of a dataset without the buffer content.The shuffle buffer will be refilled before starting to yield new examples.
{'a': 17}
{'a': 15}
{'a': 24}
{'a': 19}
{'a': 21}
{'a': 23}
...
```

where you only lose 10 rows ([1, 2, 3, 4, 5, 8, 9, 10, 11, 12])

**Comment by howitry** [comment]
> Hi ! As you can read in the logs, the shuffle buffer content is lost when resuming a shuffled dataset. The default size is 1000 examples, but you can tweak it
> 
> e.g. if you run your code with this
> 
> - ds = Dataset.from_dict({"a": range(12)}).to_iterable_dataset(num_shards=1)
> - ds = ds.shuffle(seed=42)
> + ds = Dataset.from_dict({"a": range(100)}).to_iterable_dataset(num_shards=1)
> + ds = ds.shuffle(seed=42, buffer_size=10)
> then you get
> 
> ```
> {'a': 0}
> {'a': 7}
> {'a': 6}
> checkpoint
> restart from checkpoint
> Loading a state dict of a shuffle buffer of a dataset without the buffer content.The shuffle buffer will be refilled before starting to yield new examples.
> {'a': 17}
> {'a': 15}
> {'a': 24}
> {'a': 19}
> {'a': 21}
> {'a': 23}
> ...
> ```
> 
> where you only lose 10 rows ([1, 2, 3, 4, 5, 8, 9, 10, 11, 12])

Thank you for your answer. So, when ShuffledDataSourcesArrowExamplesIterable resumes training, it will definitely discard unused data in buffer_size?

**Comment by lhoestq** [comment]
Yes correct. This is because the state_dict doesn't save the content of the buffer, so when resuming the buffer starts empty and the examples that were in the buffer are lost.

**Comment by MohammadSadeghSalehi** [comment]
The buffer loss is clear from the warning and from @lhoestq's explanation above. Separately from that, resuming can also **re-emit an example that was already yielded before the checkpoint**, which the buffer explanation does not cover: a discarded buffer accounts for missing examples, not for repeated ones.

Reproduced on `datasets` 5.0.1, CPython 3.12, macOS arm64, with no other libraries involved:

```python
from collections import Counter
from datasets import Dataset

rows = [{"id": s * 50 + i} for s in range(4) for i in range(50)]
map_ds = Dataset.from_list(rows)

def build():
    return map_ds.to_iterable_dataset(num_shards=4).shuffle(seed=17, buffer_size=16)

baseline = [r["id"] for r in build()]

ds = build()
prefix = []
for i, x in enumerate(ds):
    prefix.append(x["id"])
    if i + 1 == 20:          # k = 20
        state = ds.state_dict()
        break

ds2 = build()
ds2.load_state_dict(state)
actual = prefix + [r["id"] for r in ds2]

dups = {i: c for i, c in Counter(actual).items() if c > 1}
print(dups)                                   # {58: 2}
print({i: (i in prefix) for i in dups})       # {58: True}
print(len(set(baseline) - set(actual)))       # 17
```

Example 58 is yielded once before the checkpoint and again after restore.

Sweeping the checkpoint position k from 1 to 100 over the same dataset, four shards of fifty, seed 17:

| buffer_size | number of k values producing a duplicate | example k values |
| ---: | ---: | --- |
| 4 | 22 | many |
| 8 | 9 | |
| 16 | 3 | 20, 57, 83 |
| 32 | 0 | none in range |

Where a duplicate appears it is always exactly one id, and that id is always one already present in the prefix. Smaller buffers hit it more often. I checked that the missing count tracks `buffer_size` as expected from the documented behaviour, so this sits on top of that rather than replacing it.

Is the duplication expected as part of the same buffer refill, or is it a separate defect? It matters for training rather than for correctness of the API surface: losing roughly `buffer_size` examples on each resume is a known cost that can be reasoned about, whereas silently seeing one example twice changes the sampling distribution in a way that is invisible in a loss curve, and a job that resumes often would accumulate it.

Happy to dig further into where in the restore path the extra emission comes from if that would be useful.

**Comment by MohammadSadeghSalehi** [comment]
Following up with a narrower diagnosis, since the re-emission is separable from the buffer loss.

**Without shuffling, resume is exact.** Same dataset, same checkpoints, no shuffle:

```
plain, no shuffle        k=5   repeated=0  missing=0
plain, no shuffle        k=20  repeated=0  missing=0
plain, no shuffle        k=37  repeated=0  missing=0
plain, no shuffle        k=51  repeated=0  missing=0
plain, no shuffle        k=64  repeated=0  missing=0
```

With `shuffle(seed=17, buffer_size=16)` on the same source, `missing` is consistently
16, matching the discarded buffer as described above. But at one checkpoint an example is
also **repeated**:

```
shuffle buffer_size=16   k=5   repeated=0            missing=16
shuffle buffer_size=16   k=20  repeated=1  [58]      missing=17
shuffle buffer_size=16   k=37  repeated=0            missing=16
shuffle buffer_size=16   k=51  repeated=0            missing=16
shuffle buffer_size=16   k=64  repeated=0            missing=16
```

Id 58 is yielded before the checkpoint and again after it. Across a sweep of seeds,
checkpoint offsets, buffer sizes and shard counts it reproduces in 3 of 27
configurations, and whenever it does, the repeated example is exactly the last one
yielded before the checkpoint.

**Why.** `BufferShuffledExamplesIterable` holds no state of its own:

```python
def _init_state_dict(self) -> dict:
    self._state_dict = self.ex_iterable._init_state_dict()
```

so the checkpoint is the *source* iterable's position. That position is neither the last
example yielded nor a boundary the buffer respects: the buffer pulls ahead of what it has
yielded, yields in random order, and the source's state only commits at chunk boundaries
in `RebatchedArrowExamplesIterable`. When the last yielded example came from a chunk whose
position had not yet been committed, resuming replays that chunk and re-emits it.

So the discarded buffer explains missing examples, and this accounts for the repeated one.

Repro (datasets 5.0.1, CPython 3.12, macOS arm64):

```python
import warnings; warnings.filterwarnings("ignore")
from datasets import Dataset

rows = [{"id": s * 50 + i} for s in range(4) for i in range(50)]
map_ds = Dataset.from_list(rows)

def check(label, build, k):
    ds = build(); pre = []
    for i, x in enumerate(ds):
        pre.append(x["id"])
        if i + 1 == k:
            st = ds.state_dict(); break
    ds2 = build(); ds2.load_state_dict(st)
    suf = [r["id"] for r in ds2]
    rep = sorted(set(pre) & set(suf))
    print(f"{label:<26} k={k:<3} repeated={len(rep)} {rep[:4]} missing={200 - len(set(pre) | set(suf))}")

for k in (5, 20, 37, 51, 64):
    check("plain, no shuffle", lambda: map_ds.to_iterable_dataset(num_shards=4), k)
for k in (5, 20, 37, 51, 64):
    check("shuffle buffer_size=16",
          lambda: map_ds.to_iterable_dataset(num_shards=4).shuffle(seed=17, buffer_size=16), k)
```

Losing buffered examples on resume is a documented trade-off, but re-emitting one that was
already yielded is a different matter: for training that resumes often it means duplicate
gradient steps on specific examples, and it breaks the guarantee that an epoch visits each
example at most once.

Happy to open a PR if you have a preferred direction. The options I can see are persisting
the buffer content in the state dict, which is the straightforward fix but changes the
checkpoint size, or having the buffer track the source position corresponding to the last
example it actually yielded so resume never moves backwards over already-yielded data.

**Comment by thejamesgore** [comment]
Good timing, I hit the same duplicate a few days ago and went down the rabbit hole on it, so I can add a bit to your diagnosis.

The replay you spotted comes from `RebatchedArrowExamplesIterable`. It resumes by rewinding the inner iterable to its last snapshot and then skipping the chunks it already delivered, which is fine while the inner stream is deterministic. 

The shuffle buffer isn't though as I discovered it comes back empty with a fresh rng, so the skip discards different examples than the ones originally delivered and a row that was already yielded before the checkpoint can come out again later. 

I logged what the inner iterable consumes on both sides of the checkpoint and the duplicates match "rows re-read after already being yielded", including your `{58: 2}`.

I actually went with your first option already and opened #8408 4 days ago

The buffer content and rng position go into the state dict, so resume is exact and both the loss and the duplicate go away. Checkpoints stay small since the buffered arrow slices get compacted first, and your repro from this thread is in the regression tests.

**Comment by MohammadSadeghSalehi** [comment]
Thanks, that is a better account of the mechanism than mine. I had stopped at `BufferShuffledExamplesIterable` keeping no state of its own and reporting the source position; the piece I was missing is that the rewind-and-skip only holds while the inner stream replays deterministically, and the buffer breaks that by coming back empty with a fresh rng. That accounts for the checkpoint-offset dependence I could reproduce but not explain.

I ran my sweep against #8408 and it is clean: 27 configurations over seed, checkpoint offset, buffer size and shard count, no duplicates, no missing rows, and the resumed stream exactly equal to the uninterrupted one. Old checkpoints still load and stay exactly as lossy as before. Details on the PR.

No further work needed from me here, #8408 covers it.
